src/BSON/Iterator.c

```diff
@@ -0,0 +1,252 @@
+
+static bool php_phongo_iterator_init_with_zval(php_phongo_iterator_t* iterator, zval* zbson)
+{
+	php_phongo_document_t*  document_intern;
+	php_phongo_arraylist_t* arraylist_intern;
+	const bson_t*           bson;
+
+	ZVAL_COPY(&iterator->bson, zbson);
+	if (instanceof_function(Z_OBJCE_P(zbson), php_phongo_document_ce)) {
+		document_intern = Z_DOCUMENT_OBJ_P(&iterator->bson);
+		bson            = document_intern->bson;
+	} else if (instanceof_function(Z_OBJCE_P(zbson), php_phongo_arraylist_ce)) {
+		arraylist_intern   = Z_ARRAYLIST_OBJ_P(&iterator->bson);
+		bson               = arraylist_intern->bson;
+		iterator->is_array = true;
+	} else {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT, "Could not create iterator for %s instance.", PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(zbson));
+
+		return false;
+	}
+
+	if (!bson_iter_init(&iterator->iter, bson)) {
+		phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "Could not create iterator for BSON instance.");
+
+		return false;
+	}
+
+	iterator->valid = bson_iter_next(&iterator->iter);
+
+	return true;
+}
+
+static void php_phongo_iterator_free_current(php_phongo_iterator_t* intern)
+{
+	if (Z_ISUNDEF(intern->current)) {
+		return;
+	}
+
+	zval_ptr_dtor(&intern->current);
+	ZVAL_UNDEF(&intern->current);
+}
+
+static void php_phongo_iterator_build_current(php_phongo_iterator_t* intern)
+{
+	php_phongo_iterator_free_current(intern);
+	php_phongo_bson_iter_to_zval(&intern->current, &intern->iter);
+}
+
+static HashTable* php_phongo_iterator_get_properties_hash(phongo_compat_object_handler_type* object, bool is_temp)
+{
+	php_phongo_iterator_t* intern;
+	HashTable*             props;
+
+	intern = Z_OBJ_ITERATOR(PHONGO_COMPAT_GET_OBJ(object));
+
+	PHONGO_GET_PROPERTY_HASH_INIT_PROPS(is_temp, intern, props, 1);
+
+	zend_hash_str_update(props, "bson", sizeof("bson") - 1, &intern->bson);
+	Z_TRY_ADDREF(intern->bson);
+
+	return props;
+}
+
+PHONGO_DISABLED_CONSTRUCTOR(MongoDB_BSON_Iterator)
+PHONGO_DISABLED_WAKEUP(MongoDB_BSON_Iterator)
+
+static PHP_METHOD(MongoDB_BSON_Iterator, current)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	if (!intern->valid) {
+		RETURN_NULL();
+	}
+
+	if (Z_ISUNDEF(intern->current)) {
+		php_phongo_iterator_build_current(intern);
+	}
+
+	if (Z_ISUNDEF(intern->current)) {
+		RETURN_NULL();
+	} else {
+		ZVAL_COPY_DEREF(return_value, &intern->current);
+	}
+}
+
+static PHP_METHOD(MongoDB_BSON_Iterator, key)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	if (!intern->valid) {
+		RETURN_NULL();
+	}
+
+	if (intern->is_array) {
+		RETURN_LONG(intern->key);
+	}
+
+	RETURN_STRING(bson_iter_key(&intern->iter));
+}
+
+static PHP_METHOD(MongoDB_BSON_Iterator, next)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern->valid = bson_iter_next(&intern->iter);
+	intern->key++;
```

Comment:
Good point. I suppose we can return `null` if the iterator is exhausted and have `next` be NOP calls.

Reply:
I'll note that any call to `key` will currently fail as I've marked the return value of `key` as non-nullable but we're returning . I've asked in Room 11 to see if there's any specified behaviour, other than `foreach` not calling `key` or `current` after `valid` returned `false`. I'll defer updates to this logic to a later point.

Reply:
The `foreach` behavior is only a model for how userland should iterate manaully.

I didn't realize that the return type hint would catch this. Given that, I think it's definitely more reasonable to throw than rely on PHP for emit a typing error.

Reply:
I added a test that verifies that a type error is thrown. However, this test may not be sufficient as PHP 7 does not support union types, so the `key()` method does not have a return type there. I'll confirm behaviour on those versions and adjust the test and logic accordingly.

Reply:
PHP 8 has [InternalIterator](https://www.php.net/manual/en/class.internaliterator.php), which we obviously can't use. That was something I remember looking at when implementing Iterator on the Cursor class ([PHPC-1691](https://jira.mongodb.org/browse/PHPC-1691)).

I don't think we want to maintain separate iterator classes for each object, so I'd be easily convinced not to care about this and let PHP 7's behavior differ slightly.

Alternatively, I may as well ask: why even bother with a separate Iterator class? Could Document and Array just implement Iterator directly? I don't think the duplicated code would be that much, since Iterator.c is mostly separate overhead. This might add a few extra doc pages (mostly copy pasted), but it'd also be easier to read and require less jumping around.

With respect to serialization, we could still punt on handling the iterator offset for now.

Reply:
The main argument for implementing `IteratorAggregate` was to allow multiple iterators on the same BSON structure. If the classes were to implement `Iterator` directly, we would always be limited to a single iterator for each BSON struct. Either way, I believe this wouldn't change the issue we are having with undefined offsets vs. explicit `null` values.

Reply:
Oh, good point about allowing multiple iterators. I didn't consider that.

My thinking was that the `key()` method for separate Iterator implementations wouldn't have to use a union return type hint (Document would use `string` and ArrayList would use `int`) and thus could be enforced consistently on PHP 7.x.